### PR TITLE
Dynamic Feature Registry 

### PR DESCRIPTION
**Pull request by Eli-Liebman**
This PR adds a `FeatureRegistry` subclass which allows features to be added and managed dynamically. A use case would be adding features from a remote source, where we may not know what features flags are available at compile time.

## Description

I subclassed `FeatureRegistry` because it allowed me to accesses the `featureStore` property and was plug and play with the UI provided by the SDK.

The `DynamicFeatureRegistry` class has a `dynamicFeatures` collection which it uses to store dynamic features. When a feature is added, the backing store is checked and if an override value is found, it is applied to the feature.

A known drawback is that if a feature is added, it won't be retrieved or visible in the UI on subsequent launches unless it is re-added. This also means that there isn't a way to prune the backing store. I don't see this is a major issue but feedback would be great.

## Motivation and Context

We wanted to use YMOverride to control features flags that were part of an external SDK, without creating an internal mapping to each flag. The SDK provides a list of features that are currently available and this API would allow us to create new features at runtime.

## How Has This Been Tested?

I tested this in our application as well as in the example project.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

## License
I confirm that this contribution is made under an MIT license and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: Source/DynamicFeatureRegistry.swift [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+/// DynamicFeatureRegistry a subclass of FeatureRegistry which provides an interface for using
+/// a collection of dynamic features instead of just pre-defined ones. When a feature is added,
+/// there is an attempt to apply an override state from the underlying data FeatureStore.
+@objc open class DynamicFeatureRegistry: FeatureRegistry {
+
+    /// A local store of the current dynamic features
+    private var dynamicFeatures: [String : AnyFeature] = [:]
+
+    override var features: [LabeledItem] {
+        let dynamicLabeledItems = dynamicFeatures.values.reduce(into: []) { (result, feature) in
```
Comment (adamkaplan) [inline Source/DynamicFeatureRegistry.swift]: Why not use `map` here?

File: Source/DynamicFeatureRegistry.swift [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+            result.append(LabeledFeatureItem(label: feature.key, feature: feature))
+        }
+        let staticFeatures = super.features
+        /// merge both dynamic and static features
+        let allFeatures = dynamicLabeledItems + staticFeatures
```
Comment (adamkaplan) [inline Source/DynamicFeatureRegistry.swift]: I did read the docs at the top, but it seems arbitrary to merge static features _into_ dynamic features rather than the other way around. Should this use-case specific assumption be delegated to the user?

Perhaps something akin to Swift Dictionary's merge/merging into which provides a `uniquing:` block

File: Source/DynamicFeatureRegistry.swift [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+        // Store feature in `dynamicFeatures` if needed.
+        if dynamicFeatures[feature.key] == nil {
+            dynamicFeatures[feature.key] = feature
+        }
+
+        if let baseFeature = feature as? BaseFeature {
+            baseFeature.overrideChangeHandler = featureDidChangeHandler
+            baseFeature.overrideChangeHandler = {[unowned self] feature in
+                self.featureStore[feature.key] = feature.override
+            }
```
Comment (adamkaplan) [inline Source/DynamicFeatureRegistry.swift]: tl;dr: `weak self` is probably the way to go here.

Per [the comment](https://github.com/yahoo/Override/pull/27/files#diff-911730d490d6624dd03e9e6c31620d95b8f0b87632257cbca0e4fabd61dad159R75-R80) in FeatureRegistry below (L75-80), the `unowned` keyword was chosen because the corresponding object's lifecycle was provably tied to the lifecycle of `self`. I'm not so sure if that is true here because `feature` is currently owned by the caller. Although line 35 stores `feature` in `self.dynamicFeatures`, taking some ownership, the line 34 if-check might cause that to be bypassed resulting in a break in this chain.

In other words, if two callers pass two different `feature` parameters that have equivalent (as in ==) values in `feature.key`, one of those `feature` params will be mutated such that `overrideChangeHandler` contains a block with unowned self, but that instance will have no linkage to the lifecycle of self.

File: Source/DynamicFeatureRegistry.swift [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        let allFeatures = dynamicLabeledItems + staticFeatures
+        return allFeatures
+    }
+
+    /// This method is used to configure and store a dynamic feature. It will attempt to
+    /// assign an `override` value from the FeatureStore if it was previously set and then stores it
+    /// locally for future use.
+    /// - Parameter feature: A new feature
+    /// - Returns: The configured feature
+    public func configureDynamicFeature(feature: AnyFeature) -> AnyFeature {
```
Comment (adamkaplan) [inline Source/DynamicFeatureRegistry.swift]: Can you rename this to `addFeature()` or `add(feature:)`?

File: Source/DynamicFeatureRegistry.swift [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        // Bootstrap the last saved override value, if any
+        feature.override = featureStore[feature.key]
+
+        // Store feature in `dynamicFeatures` if needed.
+        if dynamicFeatures[feature.key] == nil {
```
Comment (adamkaplan) [inline Source/DynamicFeatureRegistry.swift]: If a similar feature instance was added previously, this method mutates the incoming feature and returns it without saving. That feels like a design flaw because it potentially breaks objects that it doesn't own and the caller can't tell if it was successful. I don't have a perfect solution to propose but I suugest replacing this check with a `guard` statement at the start of the method. If `feature.key` is already present in `dynamicFeatures`, it can take one of these paths:

 1. return that previous instance. (and document that)
 2. throw an exception.
 3. return nil.

Which one is appropriate depends on what you had in mind for the use case where multiple independent SDKs might be using the same registry instance.

**Review by adamkaplan — CHANGES_REQUESTED** [review]
Some code comments and a few philosophical ones

File: Source/DynamicFeatureRegistry.swift (hunk 5 above)
Comment (Eli-Liebman) [reply Source/DynamicFeatureRegistry.swift]: Really good points.

I think mutating and returning the same feature wasn't great. It was nice to consume but as you pointed out, the side effects were unclear. Additionally, this method didn't address what would happen if a dynamic feature conflicted with a statically defined one. 

I removed the return and added a simple getter method instead. It will be another line at the call site, assuming you want the enabled state of the feature immediately, but I think it's better overall.

File: Source/DynamicFeatureRegistry.swift (hunk 4 above)
Comment (Eli-Liebman) [reply Source/DynamicFeatureRegistry.swift]: done

File: Source/DynamicFeatureRegistry.swift (hunk 3 above)
Comment (Eli-Liebman) [reply Source/DynamicFeatureRegistry.swift]: Got it 👍

File: Source/DynamicFeatureRegistry.swift (hunk 1 above)
Comment (Eli-Liebman) [reply Source/DynamicFeatureRegistry.swift]: Good question - I feel like there was a reason but whatever it was, it no longer relevant.

File: Source/DynamicFeatureRegistry.swift (hunk 2 above)
Comment (Eli-Liebman) [reply Source/DynamicFeatureRegistry.swift]: I'm combining the two data sets so both can be displayed and managed in the UI - Is the suggestion here to do `staticFeatures + dynamicLabeledItems` instead?

Comment (dbgrandi) [reply Source/DynamicFeatureRegistry.swift]: Not sure if there is one right way to do it, so Adam's suggestion of a closure to handle the merge is interesting. (ala https://developer.apple.com/documentation/swift/dictionary/3127171-merge)

An example of where I could see this being useful is to report name conflicts on that merge with instrumentation events, even if we do end up giving priority to one set.

**Comment by dbgrandi** [comment]
@Eli-Liebman, I love the concept here and think it could be quite useful.

You mention your use case as an SDK that can register a number of features at runtime. Would it make sense to support FeatureGroups so that an SDK can create a sort of namespace that holds all their features?  I realize that may really increase the scope of this PR, but am wondering if that would result in a cleaner integration point for SDKs and also a cleaner UI in the `FeaturesTableViewController`.

File: Source/DynamicFeatureRegistry.swift [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    /// - Throws: If the provided feature conflicts with an already defined static feature, an exemption will be thrown and
+    /// it will not be added.
+    public func add(feature: AnyFeature, forced: Bool = false) throws {
+
+        /// We can't add a dynamic feature which conflicts which conflicts with a static one.
+        guard !(super.features.contains(where: { return $0.label == feature.key })) else {
+            throw DynamicFeatureRegistryError.staticFeatureAlreadyExists
+        }
+        /// If the feature is already stored and `forced` is `false`, there is no reason to contine
+        guard dynamicFeatures[feature.key] == nil || forced else { return }
```
Comment (adamkaplan) [inline Source/DynamicFeatureRegistry.swift]: In the use case from the PR description, wouldn't it also be possible for the external SDK or remote source to provide duplicate named features? I'm trying to think how to eliminate any leakage between internal Override rules and external systems behaviors.

I'd say, also just throw an exception in this. That seem reasonable?

**Review by adamkaplan — COMMENTED** [review]
Thanks for addressing the comments so quickly. I added one question about the error handling otherwise it looks good.

(I'm also trying to think of ways to handle the merging more gracefully per the first review comments)

**Comment by adamkaplan** [comment]
> @Eli-Liebman, I love the concept here and think it could be quite useful.
> 
> You mention your use case as an SDK that can register a number of features at runtime. Would it make sense to support FeatureGroups so that an SDK can create a sort of namespace that holds all their features? I realize that may really increase the scope of this PR, but am wondering if that would result in a cleaner integration point for SDKs and also a cleaner UI in the `FeaturesTableViewController`.

It can also be done in followup PR and we can wrap this one up.

File: Source/DynamicFeatureRegistry.swift [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+@objc open class DynamicFeatureRegistry: FeatureRegistry {
+
+    /// A local store of the current dynamic features
+    private var dynamicFeatures: [String : AnyFeature] = [:]
+
+    override var features: [LabeledItem] {
+
+        let dynamicLabeledItems = dynamicFeatures.values.map {
+            LabeledFeatureItem(label: $0.key, feature: $0)
+        }
```
Comment (adamkaplan) [inline Source/DynamicFeatureRegistry.swift:51]: This might be a bug – the order is lost here when using `values` of a dictionary. It would be easy enough to keep a list `[LabeledItem]` here with the side effect of making the feature lookup method slower

Comment (Eli-Liebman) [reply Source/DynamicFeatureRegistry.swift:51]: Just to clarify, are you suggesting I simply change the type of the `dynamicLabeledItems` property to an array? 

Interestingly enough, the documentation mentions that the order of key-value pairs in a dictionary is stable between mutations but is otherwise unpredictable.

Now that I'm thinking about order, as they are added may not be the most desirable, especially if its being done dynamically. Maybe I'll provide a user customizable sort block instead. 

Edit: So exposing a sort block for the `[LabeledItem]` internal type is going to be a bit complex (unless we make it public). 

There is also nuance between static and dynamic features that I'll have to think about. I'm ok with simply sorting the dynamic features alphabetically because that's really the only way to guarantee consistency. But merging the static features and applying the same sort to them would be a direct departure from the current behavior, which retains the order in which they are defined as properties.

An option is to *not* merge the features and to instead build out the TableView UI to handle multiple sections. The first section would be the static features, with the current sort behavior, and the second would be dynamic features, sorted alphabetically.

Thoughts?

Comment (adamkaplan) [reply Source/DynamicFeatureRegistry.swift:51]: Oh yeah, it's complicated for sure. I think alpha ordering is perfectly reasonable.

(Dictionary ordering is fairly predictable as yo said, at least for the same set of data. But it's effectively random for this use case.)

**Review by adamkaplan — APPROVED** [review]
